Email Workers local development (#21399)
cloudflare/cloudflare-docs: src/content/docs/email-routing/email-workers/local-development.mdx

---
title: Local Development
sidebar:
  order: 6
  badge: Beta

---

import { Render, Type, MetaInfo, WranglerConfig } from "~/components";

You can test the behavior of an Email Worker script in local development using [wrangler dev](/workers/wrangler/commands/#dev).

This is the minimal wrangler configuration required to run an Email Worker locally:

<WranglerConfig>

```jsonc
{
  "send_email": [
    {
      "name": "EMAIL"
    }
  ]
}
```

</WranglerConfig>

:::note

If you want to deploy your script you need to [enable Email Routing](/email-routing/get-started/enable-email-routing/) and have at least one verified [destination address](/email-routing/setup/email-routing-addresses/#destination-addresses).

:::

You can now test receiving, replying and sending emails in your local environment.

> ## Receiving an email

Reviewer: ```suggestion
## Receive an email
```
We usually avoid "ing" forms in headings.

Message-ID: <6114391943504294873000@ZSH-GHOSTTY>

Hi there'
```

This is what you get in the console:

```json
{
  headers: [
    {
      key: 'received',
      value: 'from smtp.example.com (127.0.0.1) by cloudflare-email.com (unknown) id 4fwwffRXOpyR for <recipient@example.com>; Tue, 27 Aug 2024 15:50:20 +0000'
    },
    { key: 'from', value: '"John" <sender@example.com>' },
    { key: 'reply-to', value: 'sender@example.com' },
    { key: 'to', value: 'recipient@example.com' },
    { key: 'subject', value: 'Testing Email Workers Local Dev' },
    { key: 'content-type', value: 'text/html; charset="windows-1252"' },
    { key: 'x-mailer', value: 'Curl' },
    { key: 'date', value: 'Tue, 27 Aug 2024 08:49:44 -0700' },
    {
      key: 'message-id',
      value: '<6114391943504294873000@ZSH-GHOSTTY>'
    }
  ],
  from: { address: 'sender@example.com', name: 'John' },
  to: [ { address: 'recipient@example.com', name: '' } ],
  replyTo: [ { address: 'sender@example.com', name: '' } ],
  subject: 'Testing Email Workers Local Dev',
  messageId: '<6114391943504294873000@ZSH-GHOSTTY>',
  date: '2024-08-27T15:49:44.000Z',
  html: 'Hi there\n',
  attachments: []
}
```

> ## Sending an email

Reviewer: ```suggestion
## Send an email
```

> Wrangler can also simulate sending emails locally. Consider this example Email Worker script that uses the [mimetext](https://www.npmjs.com/package/mimetext) npm package:

Reviewer: ```suggestion
Wrangler can also simulate sending emails locally. Consider this example Email Worker script that uses the [`mimetext`](https://www.npmjs.com/package/mimetext) npm package:
```

```ts
import { EmailMessage } from "cloudflare:email";
import { createMimeMessage } from 'mimetext';

export default {
	async fetch(request, env, ctx) {
		const msg = createMimeMessage();
		msg.setSender({ name: 'Sending email test', addr: 'sender@example.com' });
		msg.setRecipient('recipient@example.com');
		msg.setSubject('An email generated in a worker');
		msg.addMessage({
			contentType: 'text/plain',
			data: `Congratulations, you just sent an email from a worker.`,
		});

		var message = new EmailMessage('sender@example.com', 'recipient@example.com', msg.asRaw());
		await env.EMAIL.send(message);
		return Response.json({ ok: true });
	}
};
```

> Now when you run `npx wrangler dev`, go to http://localhost:8787/ to trigger the `fetch()` handler and send the email. You will see the follow message in your terminal:

Reviewer: ```suggestion
Now when you run `npx wrangler dev`, go to http://localhost:8787/ to trigger the `fetch()` handler and send the email. Your terminal will show the following:
```
Let's avoid "see" here.

> ```bash

Reviewer: ```suggestion
```txt
```
Let's use plain text since these are not `bash` commands.

```ts
import * as PostalMime from 'postal-mime';
import { createMimeMessage } from 'mimetext';
import { EmailMessage } from 'cloudflare:email';

export default {
	async email(message, env: any, ctx: any) {
		// parses incoming message
		const parser = new PostalMime.default();
		const rawEmail = new Response(message.raw);
		const email = await parser.parse(await rawEmail.arrayBuffer());

		// creates some ticket
		// const ticket = await createTicket(email);

		// creates reply message
		const msg = createMimeMessage();
		msg.setSender({ name: 'Thank you for your contact', addr: 'sender@example.com' });
		msg.setRecipient(message.from);
		msg.setHeader('In-Reply-To', message.headers.get('Message-ID'));
		msg.setSubject('An email generated in a worker');
		msg.addMessage({
			contentType: 'text/plain',
			data: `This is an automated reply. We received you email with the subject "${email.subject}", and will handle it as soon as possible.`,
		});

		const replyMessage = new EmailMessage('sender@example.com', message.from, msg.asRaw());

		await message.reply(replyMessage);
		await message.forward("recipient@example.com");
	},
};
```

> Run `npx wrangler dev` and use curl to POST the same message from the [Receiving an email](#receiving-an-email) example. Your terminal will show you where to find the replied message in your local disk and to whom the email was forwarded:

Reviewer: ```suggestion
Run `npx wrangler dev` and use curl to `POST` the same message from the [Receive an email](#receive-an-email) example. Your terminal will show you where to find the replied message in your local disk and to whom the email was forwarded:
```

> ```bash

Reviewer: ```suggestion
```txt
```

		await env.EMAIL.send(message);
		return Response.json({ ok: true });
	}
};
```

Now when you run `npx wrangler dev`, go to http://localhost:8787/ to trigger the `fetch()` handler and send the email. You will see the follow message in your terminal:

```bash
⎔ Starting local server...
[wrangler:inf] Ready on http://localhost:8787
[wrangler:inf] GET / 200 OK (19ms)
[wrangler:inf] send_email binding called with the following message:
  /var/folders/33/pn86qymd0w50htvsjp93rys40000gn/T/miniflare-f9be031ff417b2e67f2ac4cf94cb1b40/files/email/33e0a255-a7df-4f40-b712-0291806ed2b3.eml
```

Wrangler simulated `env.EMAIL.send()` by writing the email to a local file in [eml](https://datatracker.ietf.org/doc/html/rfc5322) format. The file contains the raw email message:

```
Date: Fri, 04 Apr 2025 12:27:08 +0000
From: =?utf-8?B?U2VuZGluZyBlbWFpbCB0ZXN0?= <sender@example.com>
To: <recipient@example.com>
Message-ID: <2s95plkazox@example.com>
Subject: =?utf-8?B?QW4gZW1haWwgZ2VuZXJhdGVkIGluIGEgd29ya2Vy?=
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 7bit

Congratulations, you just sent an email from a worker.
```

> ## Replying to and forwarding messages

Reviewer: ```suggestion
## Reply to and forward messages
```

---
title: Local Development
sidebar:
  order: 6
  badge: Beta

---

import { Render, Type, MetaInfo, WranglerConfig } from "~/components";

You can test the behavior of your `email()` handler in local development using [wrangler dev](/workers/wrangler/commands/#dev).

Right now local development for Email Workers is in beta. To use it, you need to install a preview release of Wrangler.

```bash
npm install --save-dev https://prerelease-registry.devprod.cloudflare.dev/workers-sdk/runs/14106373814/npm-package-wrangler-8375
```

> Make sure you have Email Routing [enabled](/email-routing/get-started/enable-email-routing/), at least one verified [destination address](/email-routing/setup/email-routing-addresses/#destination-addresses), and your wrangler configuration declares the Email Workers binding:

Reviewer 2: Enabling email routing isn't needed until you actually deploy

Author: fixed.

<WranglerConfig>

```jsonc
{

> 	"dev": {
> 		"port": 8787,
>     "local_protocol": "http"
> 	},

Reviewer 2: ```suggestion
```

  "send_email": [
    {
      "name": "EMAIL"
    }
  ]
}
```

</WranglerConfig>

You can now test receiving, replying and sending emails in your local environment.

## Receiving an email

Consider this example Email Worker script that uses the opensource [postal-mime](https://www.npmjs.com/package/postal-mime) email parser:

> ```ts
> import * as PostalMime from 'postal-mime';
> 
> export default {
> 
> 	async email(message, env: any, ctx: any) {
> 		const parser = new PostalMime.default();
> 		const rawEmail = new Response(message.raw);
> 		const email = await parser.parse(await rawEmail.arrayBuffer());
> 		console.log(email);
> 	},
> 
> };
> ```

Reviewer 2: ```suggestion
```ts
import * as PostalMime from 'postal-mime';

export default {
	async email(message, env, ctx) {
		const parser = new PostalMime.default();
		const rawEmail = new Response(message.raw);
		const email = await parser.parse(await rawEmail.arrayBuffer());
		console.log(email);
	},
};
```
```

> Now when you run `npx wrangler dev`, wrangler will will expose a local `/cdn-cgi/handler/email` endpoint that you can `POST` email messages to and trigger your Worker script `email()` handler:

Reviewer 2: ```suggestion
Now when you run `npx wrangler dev`, wrangler will expose a local `/cdn-cgi/handler/email` endpoint that you can `POST` email messages to and trigger your Worker's `email()` handler:
```

    },
    { key: 'from', value: '"John" <sender@example.com>' },
    { key: 'reply-to', value: 'sender@example.com' },
    { key: 'to', value: 'recipient@example.com' },
    { key: 'subject', value: 'Testing Email Workers Local Dev' },
    { key: 'content-type', value: 'text/html; charset="windows-1252"' },
    { key: 'x-mailer', value: 'Curl' },
    { key: 'date', value: 'Tue, 27 Aug 2024 08:49:44 -0700' },
    {
      key: 'message-id',
      value: '<6114391943504294873000@ZSH-GHOSTTY>'
    }
  ],
  from: { address: 'sender@example.com', name: 'John' },
  to: [ { address: 'recipient@example.com', name: '' } ],
  replyTo: [ { address: 'sender@example.com', name: '' } ],
  subject: 'Testing Email Workers Local Dev',
  messageId: '<6114391943504294873000@ZSH-GHOSTTY>',
  date: '2024-08-27T15:49:44.000Z',
  html: 'Hi there\n',
  attachments: []
}
```

## Sending an email

Wrangler can also simulate sending emails locally. Consider this example Email Worker script that uses the [mimetext](https://www.npmjs.com/package/mimetext) npm package:

```ts
import { EmailMessage } from "cloudflare:email";
import { createMimeMessage } from 'mimetext';

export default {

> 

Reviewer 2: ```suggestion
```

	async fetch(request, env, ctx) {
		const msg = createMimeMessage();
		msg.setSender({ name: 'Sending email test', addr: 'sender@example.com' });
		msg.setRecipient('recipient@example.com');
		msg.setSubject('An email generated in a worker');
		msg.addMessage({
			contentType: 'text/plain',
			data: `Congratulations, you just sent an email from a worker.`,
		});

		var message = new EmailMessage('sender@example.com', 'recipient@example.com', msg.asRaw());
		await env.EMAIL.send(message);
		return Response.json({ ok: true });
	}

> 

Reviewer 2: ```suggestion
```

Author: fixed

};
```

Now when you run `npx wrangler dev`, go to http://localhost:8787/ to trigger the `fetch()` handler and send the email. You will see the follow message in your terminal:

```bash
[wrangler:inf] GET / 200 OK (19ms)
send_email binding called with the following message:
/var/folders/33/pn86qymd0w50htvsjp93rys40000gn/T/miniflare-25d0146a88267bfc8d4b91b42ad5a4d0/files/28318e5d-d1c8-4bf6-b2c1-9dcb40088391.eml
```

> Wrangler simulated `env.EMAIL.send()` by writing the email to a local file in [eml](https://datatracker.ietf.org/doc/html/rfc822). The file contains the raw email message:

Reviewer 2: Should this link to the current rfc5322?

```ts
import * as PostalMime from 'postal-mime';
import { createMimeMessage } from 'mimetext';
import { EmailMessage } from 'cloudflare:email';

export default {
	async email(message, env: any, ctx: any) {
		// parses incoming message
		const parser = new PostalMime.default();
		const rawEmail = new Response(message.raw);
		const email = await parser.parse(await rawEmail.arrayBuffer());

		// creates some ticket
		// const ticket = await createTicket(email);

		// creates reply message
		const msg = createMimeMessage();
		msg.setSender({ name: 'Thank you for your contact', addr: 'sender@example.com' });
		msg.setRecipient(message.from);
		msg.setHeader('In-Reply-To', message.headers.get('Message-ID'));
		msg.setSubject('An email generated in a worker');
		msg.addMessage({
			contentType: 'text/plain',
			data: `This is an automated reply. We received you email with the subject "${email.subject}", and will handle it as soon as possible.`,
		});

> 		var replyMessage = new EmailMessage('sender@example.com', message.from, msg.asRaw());

Reviewer 2: ```suggestion
		const replyMessage = new EmailMessage('sender@example.com', message.from, msg.asRaw());
```

		await message.reply(replyMessage);
		await message.forward("recipient@example.com");
	},
};
```

Run `npx wrangler dev` and use curl to POST the same message from the [Receiving an email](#receiving-an-email) example. Your terminal will show you where to find the replied message in your local disk and to whom the email was forwarded:

```bash
[wrangler:inf] Ready on http://localhost:8787
⎔ Starting local server...
.reply() called from Email Handler with the following message:
/var/folders/33/pn86qymd0w50htvsjp93rys40000gn/T/miniflare-9deb89ed3538650182911d0c23676206/files/935d333b-3850-4d3b-b042-4d3b9b53469c.eml
.forward() called from Email Handler with

> rcptTo: recipient@example.com

Reviewer 2: This log format has changed slightly in the latest version

Author: fixed
